File: scripts/verify_webhooks/verify.js
#!/usr/bin/env node
// Verifies Tally-Signature per docs/03 §4 against every golden vector.
// Usage: node verify.js [path/to/signature-vectors.json]
const fs = require("fs");
const path = require("path");
const crypto = require("crypto");

function sign(secret, t, body) {
  return crypto.createHmac("sha256", secret).update(`${t}.${body}`).digest("hex");
}

// Real receiver logic: parse `t=<ts>,v1=<hex>[,v1=...]`, reject if now-t > tolerance,
// accept if ANY v1 matches ANY secret currently valid for this receiver (constant-time compare).
function classify(header, rawBody, secrets, now, toleranceSeconds) {
  const parts = header.split(",");
  const t = parseInt(parts[0].slice(2), 10);
  const sigs = parts.slice(1).map((p) => p.slice(3));
  if (now - t > toleranceSeconds) return "REJECT_TIMESTAMP";
  const ok = secrets.some((secret) => {
    const expected = Buffer.from(sign(secret, t, rawBody));
    return sigs.some((sig) => {
      const got = Buffer.from(sig);
      return got.length === expected.length && crypto.timingSafeEqual(got, expected);
    });
  });
  return ok ? "VALID" : "REJECT_SIGNATURE";
}

function main() {
  const file = process.argv[2] || path.join(__dirname, "..", "..", "contracts", "webhooks", "signature-vectors.json");
  const doc = JSON.parse(fs.readFileSync(file, "utf8"));
  for (const v of doc.vectors) {
    const got = classify(v.header, v.raw_body, v.secrets_valid_for_receiver, v.now, doc.tolerance_seconds);
    if (got !== v.expected) {
      console.error(`FAIL ${v.name}: got ${got}, expected ${v.expected}`);
      process.exit(1);
    }
    console.log(`ok ${v.name} -> ${got}`);
  }
  console.log(`${doc.vectors.length} vectors verified (node)`);
}

main();

File: scripts/verify_webhooks/verify.py
#!/usr/bin/env python3
"""Verifies Tally-Signature per docs/03 §4 against every golden vector.
Usage: python verify.py [path/to/signature-vectors.json]"""
import hmac
import hashlib
import json
import sys
from pathlib import Path


def sign(secret: str, t: int, body: str) -> str:
    return hmac.new(secret.encode(), f"{t}.{body}".encode(), hashlib.sha256).hexdigest()


def classify(header: str, raw_body: str, secrets: list[str], now: int, tolerance_seconds: int) -> str:
    parts = header.split(",")
    t = int(parts[0][2:])
    sigs = [p[3:] for p in parts[1:]]
    if now - t > tolerance_seconds:
        return "REJECT_TIMESTAMP"
    ok = any(hmac.compare_digest(sign(secret, t, raw_body), sig) for secret in secrets for sig in sigs)
    return "VALID" if ok else "REJECT_SIGNATURE"


def main() -> None:
    default = Path(__file__).resolve().parents[2] / "contracts" / "webhooks" / "signature-vectors.json"
    file = Path(sys.argv[1]) if len(sys.argv) > 1 else default
    doc = json.loads(file.read_text())
    for v in doc["vectors"]:
        got = classify(v["header"], v["raw_body"], v["secrets_valid_for_receiver"], v["now"], doc["tolerance_seconds"])
        if got != v["expected"]:
            print(f"FAIL {v['name']}: got {got}, expected {v['expected']}", file=sys.stderr)
            sys.exit(1)
        print(f"ok {v['name']} -> {got}")
    print(f"{len(doc['vectors'])} vectors verified (python)")


if __name__ == "__main__":
    main()
